### ISSUE TITLE
(pipelines): support caching in simple synth action

### ISSUE THREAD

**Issue by johnameyer** [aws/aws-cdk#13043]
Allow for [caching](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-codebuild-readme.html#caching) in [SimpleSynthAction](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_pipelines.SimpleSynthAction.html)

### Use Case

Allowing for caching in the SimpleSynthAction makes it easy to decrease build times - e.g. caching incremental typescript lambda builds, caching npm packages, or even re-use unchanged resources - which can become quite incredible when using CI/CD like CDK Pipelines.

### Proposed Solution

This would simply add a couple additional optional fields (`cache` and `cachePaths`) onto the SimpleSynthAction static methods, which could be used like:

```typescript
SimpleSynthAction.standardNpmSynth({
  sourceArtifact,
  cloudAssemblyArtifact,
  installCommand: 'npm ci --cache .npm',
  buildCommand: "npm run ci-build",
  cachePaths: ['dist/**/*', '.npm/**/*', './.git/objects/**/*', '.git-hash'],
  cache: Cache.bucket(new Bucket(this, 'Cache'))
});
```

Currently, you can either create the whole build configuration yourself or copy and modify the source of the SimpleSynthAction, but these both seem untenable considering how easy it is to pass these options to the constructed options.

### Other

These parameters just get passed along into the SimpleSynthAction without any direct transformation, so this is really simple:

```typescript
const buildSpec = BuildSpec.fromObject({
  ....
  cache: {
    paths: this.props.cachePaths
  }
});
```

and

```typescript
const project = new PipelineProject(scope, 'CdkBuildProject', {
  ...
  cache: this.props.cache
});
```

[SimpleSynthAction docs](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_pipelines.SimpleSynthAction.html)
[SimpleSynthAction source](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/pipelines/lib/synths/simple-synth-action.ts)
[CodeBuild caching](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-codebuild-readme.html#caching)

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by Pythoner6** [comment]
I've got an application where this would be really helpful - it has a bunch of rust-based dockerized lambdas, which building from scratch can take 10 minutes, and most of that time is compiling all the dependencies from source. It would be really nice to be able to have those bits cached to cut out a substantial portion of that build time.

**Comment by gshpychka** [comment]
@rix0rrr can you shed some light on the plans for this?

**Comment by simlevesque** [comment]
This is super important for us.

**Comment by hoegertn** [comment]
With the new API you can provide a partialBuildSpec to configure caching. The SimpleSynthAction is now part of the legacy API.

**Comment by simlevesque** [comment]
> With the new API you can provide a partialBuildSpec to configure caching. The SimpleSynthAction is now part of the legacy API.

@hoegertn 

Yes, I use a partialBuildSpec which configures caching for some folders. But there is no persistence without a project cache like the OP is saying. Right now there are no way to share a yarn/npm cache between different pipelines invocation if I'm not mistaken.

**Comment by johnameyer** [comment]
Seems like https://github.com/aws/aws-cdk/issues/16375 tracks this issue for the new version of CDK pipelines
